smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java

```diff
@@ -570,37 +584,120 @@ private void parseStructuredShape(
+        expect('w');
+        expect('i');
+        expect('t');
+        expect('h');
+
+        if (!modelFile.getVersion().supportsMixins()) {
+            throw syntax(id, "Mixins can only be used with Smithy version 2 or later. "
+                    + "Attempted to use mixins with version `" + modelFile.getVersion() + "`.");
         }
 
-        parseMembers(id, Collections.emptySet(), structureMember);
+        ws();
+        do {
+            String target = ParserUtils.parseShapeId(this);
+            modelFile.addForwardReference(target, resolved -> modelFile.addPendingMixin(id, resolved));
+            ws();
+        } while (peek() != '{');
     }
 
     private void parseOperationStatement(ShapeId id, SourceLocation location) {
-        ws();
         OperationShape.Builder builder = OperationShape.builder().id(id).source(location);
-        ObjectNode node = IdlNodeParser.parseObjectNode(this, id.toString());
-        LoaderUtils.checkForAdditionalProperties(node, id, OPERATION_PROPERTY_NAMES, modelFile.events());
+        parseProperties(id, OPERATION_PROPERTY_NAMES, propertyName -> {
+            switch (propertyName) {
+                case "input":
+                    parseInlineableOperationMember(id, operationInputSuffix, builder::input);
+                    break;
+                case "output":
+                    parseInlineableOperationMember(id, operationOutputSuffix, builder::output);
+                    break;
+                case "errors":
+                    parseIdList(builder::addError);
+                    break;
+                default:
+                    break;
+            }
+        });
+        modelFile.onShape(builder);
+    }
+
+    private void parseProperties(ShapeId id, Collection<String> allowedProperties, Consumer<String> valueParser) {
+        ws();
+        expect('{');
+        ws();
+
+        Set<String> defined = new TreeSet<>();
+        while (!eof() && peek() != '}') {
+            String key = IdlNodeParser.parseNodeObjectKey(this);
+            if (defined.contains(key)) {
+                throw syntax(id, String.format("Duplicate %s binding for %s", key, id));
+            } else if (!allowedProperties.contains(key)) {
+                throw syntax(id, String.format("Unknown property %s for %s", key, id));
+            }
+            defined.add(key);
+
+            ws();
+            expect(':');
+            valueParser.accept(key);
+            ws();
+        }
+
+        expect('}');
+    }
+
+    private void parseInlineableOperationMember(ShapeId id, String suffix, Consumer<ShapeId> consumer) {
+        if (peek() == '=') {
+            if (!modelFile.getVersion().supportsInlineOperationIO()) {
+                throw syntax(id, "Inlined operation inputs and outputs can only be used with Smithy version 2 or "
+                        + "later. Attempted to use inlined IO with version `" + modelFile.getVersion() + "`.");
+            }
+            expect('=');
```

Comment:
I think you need to clear pending docs here. For example, what happens right now with this?

```
operation Foo {
    input 
    /// These should not be part of the shape!
    := {}
}
```